### ISSUE TITLE
Feature Request / Workaround for Variable size multi-label candidate sampling in TensorFlow.

### ISSUE THREAD

**Issue by MtDersvan** [tensorflow/tensorflow#12968]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: +
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS/Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: both
- **TensorFlow version (use command below)**:  ('v1.3.0-rc2-20-g0787eee', '1.3.0')
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: n/a

### Context:
Suppose we have a dataset with an arbitrary amount of labels per each training example (image segmentation, multi-label classification, etc.). Labels (classes) are NOT mutually exclusive, thus vary in size between examples. 

### Problem:
When trying to use provided standard `nce_loss()` -  a static `int` value for `num_true` option is required. 
- `num_true`: An `int`. The number of target classes per training example.

This probably works well for problems where we have same amount of labels per training examples and we know them in advance.
When labels have a variable shape `[None]`, (and in our case, they are also being batched and bucketed by bucket size with `.padded_batch()` + `.group_by_window()`) it is necessary to supply a variable size `num_true` in order to accustom for all training examples. This is currently unsupported to my knowledge (correct me if I'm wrong).

### Statement:
Is there a proper way to do this or any other workarounds? If not I would like to request a feature. If yes, I would appreciate a working example here or on stackoverflow.

Related [stackoverflow question](https://stackoverflow.com/questions/46085454/variable-size-multi-label-candidate-sampling-in-tensorflow).

Desired behaviour:
```
nce_loss = tf.nn.nce_loss(
    weights=nce_weights,
    biases=nce_biases,
    labels=labels,
    inputs=inputs,
    num_sampled=num_sampled,
    num_true=(tf.shape(labels)[-1]), # or tf.placeholder("int32", [], name="num_trve")
    num_classes=self.num_classes)
```

Also, is it possible to add support for weighted losses to `nce_loss()` (specifically to `_compute_sampled_logits()` as it is partially generated from a .cc) in the same fashion as it is implemented in `tf.losses.sparse_softmax_cross_entropy` or `tf.losses.sigmoid_cross_entropy`?
Thanks in advance.

**Comment by cy89** [comment]
@drpngx, do you know of any workarounds for @MtDersvan 's shape problem?

**Comment by drpngx** [comment]
Sorry for the delay. What happens when you feed it an unknown shape `num_true`? Do you have a short example / stack trace you could put on pastebin? Thanks.

**Comment by MtDersvan** [comment]
@drpngx Sorry for the delay, too.
Here is a simple short toy example: https://pastebin.com/bNH8F4K5
Here is a stack trace: https://pastebin.com/H9fctCxt
Basically there's a hardcoded `int(v)` in `_MakeInt` in `log_uniform_candidate_sampler` hard type restriction. And `_log_uniform_candidate_sampler` code is generated from a kernel which makes it harder to hack it on my own.

If I understand `num_true` correctly (maybe I'm wrong?), but it should be a common case where the number of labels differs per example. Maybe there is a different sampler that supports variable `num_true`?

**Comment by MtDersvan** [comment]
I would also appreciate if someone can suggest the best way to add support for weighted losses to `nce_loss` (or maybe to `_compute_sampled_logits`?) in the same fashion as it is implemented in `tf.losses.sparse_softmax_cross_entropy` or `tf.losses.sigmoid_cross_entropy`? Because, multi-label tasks often have imbalanced datasets.

**Comment by MtDersvan** [comment]
Sorry, but is there any progress here?

**Comment by drpngx** [comment]
@MtDersvan : `num_true` is an [attr](https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/candidate_sampling_ops.cc#L100), you can't pass a tensor to it. So the right way is to have it passed as an input. Does that make sense?

**Comment by MtDersvan** [comment]
@drpngx 
I fully understand that right now you can't pass a tensor.

What do you mean as an input?
If you mean just a const number than sampling losses have a very narrow application.
If I understand correctly, in TF I can either pass a tensor or a numeric value (which is a const in itself in the context of a static graph).

Documentation:
`num_true: Number of true labels per context.` 
and
`num_true: An int. The number of target classes per training example.`

Also, If I understand correctly it is the amount of true labels per each example, isn't it? 
E.g. if a picture has ['dog', 'car', 'panda'] true labels, while there are 1000 possible classes, than `num_true=3` and `num_classes=1000`.

If yes, then what exactly needs to be done for it to support a variable size`num_true`?
E.g. if I have images with different amount of `num_true` per example.

**Comment by MtDersvan** [comment]
Oh, I see you updated the docs recently. 
```
Note: It would be useful to allow a variable number of target classes per example. We hope to 
provide this functionality in a future release. For now, if you have a variable number of target 
classes, you can pad them out to a constant number by either repeating them or by padding with 
an otherwise unused class.
```
Won't such padding interfere the results? Because of this:
```
Note: In the case where num_true > 1, we assign to each target class the target probability 1 / num_true so that the target probabilities sum to 1 per-example.
```
Do proportions/distribution matter?

It looks like you are already working on it, still if you need help implementing this feature, I would be willing to contribute.

**Comment by drpngx** [comment]
Yes, the normalizer has to know about the padding. Also, changing `num_true` to an input tensor will change the order in the parameters, so it's an interface change.

Probably the best is to create a new copy of the operator in `contrib`. I think that would be a fine contribution!

(To use a tensor parameter, simply change the `.attr` to `.input` and then change the kernel to get the input from the context each time `Compute` is requested.)

**Comment by MtDersvan** [comment]
@drpngx I think the order of parameters doesn't need to be changed, `num_true` is always the first `.attr`, so switching it to an `.input` preserves the order. I have a simple working fix that leaves almost everything intact [here](https://github.com/MtDersvan/tensorflow/tree/r1.4) on`r1.4` branch. If you're interested I can submit a PR. 
I did direct changes to `core/ops` and `core/kernels` instead of a `contrib` clone, due to the fact that it was too much of a hassle to replicate all candidate_samplers and range_samplers.
Right now ops take a single `tensor` instead of an `int` per batch.

Future Work:
I haven't figured out yet how to make it to work with a list of tensors of shape `batch_size`, so each example in a batch can have it's own `num_true`. Because rn it needs all items in a batch to have the same number of `num_true`, which is inefficient computationally. Also, I left out one check in `ComputeAccidentalHits` - `.SetShapeFn()`.

**Comment by drpngx** [comment]
Oh good, yes feel free to send the PR!

On Nov 28, 2017 12:51 PM, "MtDersvan" <notifications@github.com> wrote:

@drpngx <https://github.com/drpngx> I think the order of parameters doesn't
need to be changed, num_true is always the first .attr, so switching it to
an .input preserves the order. I have a simple working fix that leaves
almost everything intact here
<https://github.com/MtDersvan/tensorflow/tree/r1.4> onr1.4 branch. If
you're interested I can submit a PR.
I did direct changes to core/ops and core/kernels instead of a contrib
clone, due to the fact that it was too much of a hassle to replicate all
candidate_samplers and range_samplers.
Right now ops take a single tensor instead of an int per batch.

Future Work:
I haven't figured out yet how to make it to work with a list of tensors of
shape batch_size, so each example in a batch can have it's own num_true.
Because rn it needs all items in a batch to have the same number of num_true,
which is inefficient computationally. Also, I left out one check in
ComputeAccidentalHits - .SetShapeFn().

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/12968#issuecomment-347412520>,
or mute the thread

**Comment by sushreebarsa** [comment]
@MtDersvan One workaround could be to develop a custom loss function that dynamically calculates num_true based on the labels in each example. We could consider utilizing functions like tf.reduce_sum or tf.count_nonzero within your custom loss function to calculate num_true dynamically. 
Thank you!
